include/deal.II/base/index_set.h

```diff
@@ -506,14 +938,53 @@ IndexSet::Range::Range ()
 
 
 inline
-IndexSet::Range::Range (const types::global_dof_index i1,
-                        const types::global_dof_index i2)
+IndexSet::Range::Range (const size_type i1,
+                        const size_type i2)
   :
   begin(i1),
   end(i2)
 {}
 
 
+/* IndexSet itself */
+
+inline
+IndexSet::ElementIterator IndexSet::begin() const
+{
+  compress();
+  if (ranges.size()>0)
+    return IndexSet::ElementIterator(*this, 0, ranges[0].begin);
+  else
+    return IndexSet::ElementIterator(*this);
+}
+
+inline
+IndexSet::ElementIterator IndexSet::end() const
+{
+  compress();
+  if (ranges.size()>0)
+    return IndexSet::ElementIterator(*this, ranges.size()-1, ranges.back().end);
+  else
+    return IndexSet::ElementIterator(*this);
```

Comment:
You wouldn't need this if-else if you had a canonical representation for end-iterators.